File: code/104.maximum-depth-of-binary-tree.php
<?php
/*
 * @lc app=leetcode id=104 lang=php
 *
 * [104] Maximum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer
     */
    function maxDepth($root) {
        if ($root == null) return 0;

        return $this->countDeps($root, 1);
    }

    function countDeps($node, $curDepth) {
        $leftDepth = $rightDepth = $curDepth;
        if ($node->left != null) {
            $leftDepth++;
            $leftDepth = $this->countDeps($node->left, $leftDepth);
        }
        if ($node->right != null) {
            $rightDepth++;
            $rightDepth = $this->countDeps($node->right, $rightDepth);
        }

        return max($leftDepth, $rightDepth);
    }
}
// @lc code=end

File: code/104.maximum-depth-of-binary-tree.go
package code

/*
 * @lc app=leetcode id=104 lang=golang
 *
 * [104] Maximum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	return recursion(root, 1)
}

func recursion(node *TreeNode, depth int) int {
	if node == nil {
		return depth - 1
	}

	left := recursion(node.Left, depth+1)
	right := recursion(node.Right, depth+1)

	return max(left, right)
}

// @lc code=end
